Preprocessor handling

Compared to what I know from the use of icecream/icecc, I find the handling of preprocessing by sccache less optimal.

With icecream/icecc, this worked transparently. The compiler output was the same, regardless of whether the compilation was done locally or remotely.

When using sccache, the compiler output is hardly useful in case of warnings or errors, because the compilation is done on the preprocessed source code, and all macro usages have been rolled out. If the code having warnings/errors uses macros heavily, this is virtually unreadable. In addition, this yields a lot additional meaningless warnings that only occur because the compiler only sees the result of the preprocessor (such as loop executed only once, e.g.). Also, the console coloring is not applied (but this wouldn't be helpful anyway in the cases mentioned before).

I am not sure if this is true also for compilation that goes through sccache but it actually done locally, or if this only applies to compilation done remotely.

In the former case, this issue might be mitigated to some extent, by allowing to disable remote compilation for certain paths (i.e. code that the user is actively working on), or rerunning the compilation locally when remote compilation had errors.

Why does this happen? I took not of #178, which seems to imply that the behaviour was already changed to match my expectation quite some time ago, but with the current version of sccache, this still happens to me. Can this be configured?

I am seeing something like this, e.g.:
```
/home/osboxes/work/mozilla-unified/dom/indexedDB/ActorsParent.cpp:27100:1040: warning: code will never be executed [-Wunreachable-code]
MozExternalRefCountType FileHelper::ReadCallback::AddRef(void) { static_assert(!mozilla::IsDestructible<FileHelper::ReadCallback>::value, "Reference-counted class " "FileHelper::ReadCallback" " should not have a public destructor. " "Make this class's destructor non-public"); do { static_assert( mozilla::detail::AssertionConditionType<decltype(int32_t(mRefCnt) >= 0)>::isValid, "invalid assertion condition"); if ((__builtin_expect(!!(!(!!(int32_t(mRefCnt) >= 0))), 0))) { MOZ_ReportAssertionFailure("int32_t(mRefCnt) >= 0" " (" "illegal refcnt" ")", "/home/osboxes/work/mozilla-unified/dom/indexedDB/ActorsParent.cpp", 27100); AnnotateMozCrashReason("MOZ_ASSERT" "(" "int32_t(mRefCnt) >= 0" ") (" "illegal refcnt" ")"); do { *((volatile int*)__null) = 27100; ::abort(); } while (false); } } while (false); do { static_assert( mozilla::detail::AssertionConditionType<decltype("FileHelper::ReadCallback" != nullptr)>::isValid, "invalid assertion condition"); if ((__builtin_expect(!!(!(!!("FileHelper::ReadCallback" != nullptr))), 0))) { MOZ_ReportAssertionFailure("\"FileHelper::ReadCallback\" != nullptr" " (" "Must specify a name" ")", "/home/osboxes/work/mozilla-unified/dom/indexedDB/ActorsParent.cpp", 27100); AnnotateMozCrashReason("MOZ_ASSERT" "(" "\"FileHelper::ReadCallback\" != nullptr" ") (" "Must specify a name" ")"); do { *((volatile int*)__null) = 27100; ::abort(); } while (false); } } while (false); if (!mRefCnt.isThreadSafe) this->_mOwningThread.AssertOwnership("FileHelper::ReadCallback" " not thread-safe"); nsrefcnt count = ++mRefCnt; NS_LogAddRef((this), (count), ("FileHelper::ReadCallback"), (uint32_t)(sizeof(*this))); return count; };
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               ^~~~~~~~~~~~~~~~~~~~~~~~~~
```

This is a warning that wouldn't be present originally. Turning of -Wunreachable-code (and similar ones) is not a doable option, since this would also turn it off for non-preprocessing-related code locations.

Similarly, if I introduce an error on purpose, I get something like this (for some reason twice):
```
/home/osboxes/work/mozilla-unified/dom/indexedDB/ActorsParent.cpp:25735:95: error: expression is not assignable
        do { static_assert( mozilla::detail::AssertionConditionType<decltype(mResponse.type() = CursorResponse::TArrayOfObjectStoreKeyCursorResponse)>::isValid, "invalid assertion condition"); if ((__builtin_expect(!!(!(!!(mResponse.type() = CursorResponse::TArrayOfObjectStoreKeyCursorResponse))), 0))) { MOZ_ReportAssertionFailure("mResponse.type() = CursorResponse::TArrayOfObjectStoreKeyCursorResponse", "/home/osboxes/work/mozilla-unified/dom/indexedDB/ActorsParent.cpp", 25736); AnnotateMozCrashReason("MOZ_ASSERT" "(" "mResponse.type() = CursorResponse::TArrayOfObjectStoreKeyCursorResponse" ")"); do { *((volatile int*)__null) = 25736; ::abort(); } while (false); } } while (false);
                                                                             ~~~~~~~~~~~~~~~~ ^
/home/osboxes/work/mozilla-unified/dom/indexedDB/ActorsParent.cpp:25735:241: error: expression is not assignable
        do { static_assert( mozilla::detail::AssertionConditionType<decltype(mResponse.type() = CursorResponse::TArrayOfObjectStoreKeyCursorResponse)>::isValid, "invalid assertion condition"); if ((__builtin_expect(!!(!(!!(mResponse.type() = CursorResponse::TArrayOfObjectStoreKeyCursorResponse))), 0))) { MOZ_ReportAssertionFailure("mResponse.type() = CursorResponse::TArrayOfObjectStoreKeyCursorResponse", "/home/osboxes/work/mozilla-unified/dom/indexedDB/ActorsParent.cpp", 25736); AnnotateMozCrashReason("MOZ_ASSERT" "(" "mResponse.type() = CursorResponse::TArrayOfObjectStoreKeyCursorResponse" ")"); do { *((volatile int*)__null) = 25736; ::abort(); } while (false); } } while (false);
```

While in this simple case, I can actually figure out the problem from the message, it is already very hard to read. And in more complex cases (imagine involving templates), it will be even much harder.

> The alternative (running the preprocessor on the remote machine) requires setting up remote machines with exactly the same paths and exactly the same files as the local machines, which is not really doable for a lot of environments.

Doesn't icecream do exactly that automatically, i.e. package the whole environment and transport it to the remote machine, where it is cached? As I wrote above, I never saw these kind of warnings with icecream.

I didn't notice that #178 applied to local compilation only, but the problem solved by #178 (which is essentially what I reported here) applies to the remote case as well.

Ah, stupid mistake. It said `non-compilation request`, and that was right, I wasn't aware I need to do a compilation only call to have it cacheable.

Now it works as expected:
```
[simon@sigibln sccache]$ target/debug/sccache ~/.mozbuild/clang/bin/clang++ -c test.cpp -Wall -Wextra -Wunreachable-code
[simon@sigibln sccache]$ killall sccache
[simon@sigibln sccache]$ ~/.cargo/bin/sccache ~/.mozbuild/clang/bin/clang++ -c test.cpp -Wall -Wextra -Wunreachable-code
test.cpp:53:62: warning: self-comparison always evaluates to true [-Wtautological-compare]
int main() { !(((google_breakpad::LogStream::SEVERITY_ERROR) >= (google_breakpad::LogStream::SEVERITY_ERROR))) ? (void)0 : google_breakpad::LogMessageVoidify() & (std::cout) << "UTF16ToUTF8 found low surrogate "; }
                                                             ^
test.cpp:53:120: warning: code will never be executed [-Wunreachable-code]
int main() { !(((google_breakpad::LogStream::SEVERITY_ERROR) >= (google_breakpad::LogStream::SEVERITY_ERROR))) ? (void)0 : google_breakpad::LogMessageVoidify() & (std::cout) << "UTF16ToUTF8 found low surrogate "; }
                                                                                                                       ^
2 warnings generated.
```

First is my ad-hoc modified version (no warning), second is the upstream version (warning from expanded macro).

--- a/.travis.yml
+++ b/.travis.yml
@@ -33,8 +33,8 @@ before_script:
   - if [[ "${DIST_SCCACHE}" = "1" ]]; then export EXTRA_FEATURES="$EXTRA_FEATURES dist-client dist-server"; fi
 
 script:
-  - cargo build --verbose --features="all ${EXTRA_FEATURES}"
-  - RUST_BACKTRACE=1 cargo test --all --verbose --no-default-features --features="${EXTRA_FEATURES}"
+  - cargo build --verbose --features="all ${EXTRA_FEATURES}" || exit 1
+  - RUST_BACKTRACE=1 cargo test --all --verbose --no-default-features --features="${EXTRA_FEATURES}" 
   - RUST_BACKTRACE=1 cargo test --all --verbose --features="all ${EXTRA_FEATURES}"
   # Requires PR #321
   #- if [[ "${DIST_SCCACHE}" = "1" ]]; then RUST_BACKTRACE=1 cargo test --all --verbose --features="all dist-tests ${EXTRA_FEATURES}" test_dist_ -- --test-threads 1; fi

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -139,7 +139,7 @@ dist-client = ["ar", "flate2", "hyper", "hyperx", "reqwest", "url", "sha2"]
 # Enables the sccache-dist binary
 dist-server = ["crossbeam-utils", "jsonwebtoken", "flate2", "libmount", "nix", "openssl", "reqwest", "rouille", "syslog", "void", "version-compare"]
 # Enables dist tests with external requirements
-dist-tests = []
+dist-tests = ["dist-client", "dist-server"]
 
 [workspace]
 exclude = ["tests/test-crate"]

--- a/src/compiler/c.rs
+++ b/src/compiler/c.rs
@@ -173,6 +173,7 @@ pub trait CCompilerImpl: Clone + fmt::Debug + Send + 'static {
         cwd: &Path,
         env_vars: &[(OsString, OsString)],
         may_dist: bool,
+        rewrite_includes_only: bool,
     ) -> SFuture<process::Output>
     where
         T: CommandCreatorSync;
@@ -185,6 +186,7 @@ pub trait CCompilerImpl: Clone + fmt::Debug + Send + 'static {
         parsed_args: &ParsedArguments,
         cwd: &Path,
         env_vars: &[(OsString, OsString)],
+        rewrite_includes_only: bool,
     ) -> Result<(CompileCommand, Option<dist::CompileCommand>, Cacheable)>;
 }
 
@@ -250,6 +252,7 @@ where
         env_vars: Vec<(OsString, OsString)>,
         may_dist: bool,
         pool: &CpuPool,
+        rewrite_includes_only: bool,
     ) -> SFuture<HashResult> {
         let me = *self;
         let CCompilerHasher {
@@ -265,6 +268,7 @@ where
             &cwd,
             &env_vars,
             may_dist,
+            rewrite_includes_only,
         );
         let out_pretty = parsed_args.output_pretty().into_owned();
         let result = result.map_err(move |e| {
@@ -373,6 +377,7 @@ impl<I: CCompilerImpl> Compilation for CCompilation<I> {
     fn generate_compile_commands(
         &self,
         path_transformer: &mut dist::PathTransformer,
+        rewrite_includes_only: bool,
     ) -> Result<(CompileCommand, Option<dist::CompileCommand>, Cacheable)> {
         let CCompilation {
             ref parsed_args,
@@ -382,7 +387,14 @@ impl<I: CCompilerImpl> Compilation for CCompilation<I> {
             ref env_vars,
             ..
         } = *self;
-        compiler.generate_compile_commands(path_transformer, executable, parsed_args, cwd, env_vars)
+        compiler.generate_compile_commands(
+            path_transformer,
+            executable,
+            parsed_args,
+            cwd,
+            env_vars,
+            rewrite_includes_only,
+        )
     }
 
     #[cfg(feature = "dist-client")]

--- a/src/compiler/c.rs
+++ b/src/compiler/c.rs
@@ -82,7 +82,7 @@ pub struct ParsedArguments {
     pub depfile: Option<PathBuf>,
     /// Output files, keyed by a simple name, like "obj".
     pub outputs: HashMap<&'static str, PathBuf>,
-    /// Commandline arguments for the preprocessor.
+    /// Commandline arguments for the preprocessor (not including common_args).
     pub preprocessor_args: Vec<OsString>,
     /// Commandline arguments for the preprocessor or the compiler.
     pub common_args: Vec<OsString>,

--- a/src/compiler/clang.rs
+++ b/src/compiler/clang.rs
@@ -55,6 +55,7 @@ impl CCompilerImpl for Clang {
         cwd: &Path,
         env_vars: &[(OsString, OsString)],
         may_dist: bool,
+        rewrite_includes_only: bool,
     ) -> SFuture<process::Output>
     where
         T: CommandCreatorSync,
@@ -67,6 +68,7 @@ impl CCompilerImpl for Clang {
             env_vars,
             may_dist,
             self.kind(),
+            rewrite_includes_only,
         )
     }
 
@@ -77,8 +79,16 @@ impl CCompilerImpl for Clang {
         parsed_args: &ParsedArguments,
         cwd: &Path,
         env_vars: &[(OsString, OsString)],
+        rewrite_includes_only: bool,
     ) -> Result<(CompileCommand, Option<dist::CompileCommand>, Cacheable)> {
-        gcc::generate_compile_commands(path_transformer, executable, parsed_args, cwd, env_vars)
+        gcc::generate_compile_commands(
+            path_transformer,
+            executable,
+            parsed_args,
+            cwd,
+            env_vars,
+            rewrite_includes_only,
+        )
     }
 }
 

--- a/src/compiler/clang.rs
+++ b/src/compiler/clang.rs
@@ -12,7 +12,7 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-#![allow(unused_imports,dead_code,unused_variables)]
+#![allow(unused_imports, dead_code, unused_variables)]
 
 use crate::compiler::args::*;
 use crate::compiler::c::{CCompilerImpl, CCompilerKind, Language, ParsedArguments};
@@ -59,7 +59,15 @@ impl CCompilerImpl for Clang {
     where
         T: CommandCreatorSync,
     {
-        gcc::preprocess(creator, executable, parsed_args, cwd, env_vars, may_dist)
+        gcc::preprocess(
+            creator,
+            executable,
+            parsed_args,
+            cwd,
+            env_vars,
+            may_dist,
+            self.kind(),
+        )
     }
 
     fn generate_compile_commands(

--- a/src/compiler/compiler.rs
+++ b/src/compiler/compiler.rs
@@ -137,6 +137,7 @@ where
         env_vars: Vec<(OsString, OsString)>,
         may_dist: bool,
         pool: &CpuPool,
+        rewrite_includes_only: bool,
     ) -> SFuture<HashResult>;
 
     /// Return the state of any `--color` option passed to the compiler.
@@ -162,7 +163,18 @@ where
             Ok(Some(_)) => true,
             _ => false,
         };
-        let result = self.generate_hash_key(&creator, cwd.clone(), env_vars, may_dist, &pool);
+        let rewrite_includes_only = match dist_client {
+            Ok(Some(ref client)) => client.rewrite_includes_only(),
+            _ => false,
+        };
+        let result = self.generate_hash_key(
+            &creator,
+            cwd.clone(),
+            env_vars,
+            may_dist,
+            &pool,
+            rewrite_includes_only,
+        );
         Box::new(result.then(move |res| -> SFuture<_> {
             debug!(
                 "[{}]: generate_hash_key took {}",
@@ -390,7 +402,7 @@ where
 {
     let mut path_transformer = dist::PathTransformer::new();
     let compile_commands = compilation
-        .generate_compile_commands(&mut path_transformer)
+        .generate_compile_commands(&mut path_transformer, true)
         .chain_err(|| "Failed to generate compile commands");
     let (compile_cmd, _dist_compile_cmd, cacheable) = match compile_commands {
         Ok(cmds) => cmds,
@@ -420,9 +432,13 @@ where
     use futures::future;
     use std::io;
 
+    let rewrite_includes_only = match dist_client {
+        Ok(Some(ref client)) => client.rewrite_includes_only(),
+        _ => false,
+    };
     let mut path_transformer = dist::PathTransformer::new();
     let compile_commands = compilation
-        .generate_compile_commands(&mut path_transformer)
+        .generate_compile_commands(&mut path_transformer, rewrite_includes_only)
         .chain_err(|| "Failed to generate compile commands");
     let (compile_cmd, dist_compile_cmd, cacheable) = match compile_commands {
         Ok(cmds) => cmds,
@@ -585,6 +601,7 @@ pub trait Compilation {
     fn generate_compile_commands(
         &self,
         path_transformer: &mut dist::PathTransformer,
+        rewrite_includes_only: bool,
     ) -> Result<(CompileCommand, Option<dist::CompileCommand>, Cacheable)>;
 
     /// Create a function that will create the inputs used to perform a distributed compilation
@@ -1748,6 +1765,9 @@ mod test_dist {
         ) -> SFuture<(Toolchain, Option<String>)> {
             f_err("put toolchain failure")
         }
+        fn rewrite_includes_only(&self) -> bool {
+            false
+        }
     }
 
     pub struct ErrorAllocJobClient {
@@ -1789,6 +1809,9 @@ mod test_dist {
             _: Box<dyn pkg::ToolchainPackager>,
         ) -> SFuture<(Toolchain, Option<String>)> {
             f_ok((self.tc.clone(), None))
+        }
+        fn rewrite_includes_only(&self) -> bool {
+            false
         }
     }
 
@@ -1847,6 +1870,9 @@ mod test_dist {
             _: Box<dyn pkg::ToolchainPackager>,
         ) -> SFuture<(Toolchain, Option<String>)> {
             f_ok((self.tc.clone(), None))
+        }
+        fn rewrite_includes_only(&self) -> bool {
+            false
         }
     }
 
@@ -1907,6 +1933,9 @@ mod test_dist {
             _: Box<dyn pkg::ToolchainPackager>,
         ) -> SFuture<(Toolchain, Option<String>)> {
             f_ok((self.tc.clone(), Some("/overridden/compiler".to_owned())))
+        }
+        fn rewrite_includes_only(&self) -> bool {
+            false
         }
     }
 
@@ -1988,6 +2017,9 @@ mod test_dist {
             _: Box<dyn pkg::ToolchainPackager>,
         ) -> SFuture<(Toolchain, Option<String>)> {
             f_ok((self.tc.clone(), Some("/overridden/compiler".to_owned())))
+        }
+        fn rewrite_includes_only(&self) -> bool {
+            false
         }
     }
 }

--- a/src/compiler/diab.rs
+++ b/src/compiler/diab.rs
@@ -54,6 +54,7 @@ impl CCompilerImpl for Diab {
         cwd: &Path,
         env_vars: &[(OsString, OsString)],
         may_dist: bool,
+        _rewrite_includes_only: bool,
     ) -> SFuture<process::Output>
     where
         T: CommandCreatorSync,
@@ -68,6 +69,7 @@ impl CCompilerImpl for Diab {
         parsed_args: &ParsedArguments,
         cwd: &Path,
         env_vars: &[(OsString, OsString)],
+        _rewrite_includes_only: bool,
     ) -> Result<(CompileCommand, Option<dist::CompileCommand>, Cacheable)> {
         generate_compile_commands(path_transformer, executable, parsed_args, cwd, env_vars)
     }

--- a/src/compiler/gcc.rs
+++ b/src/compiler/gcc.rs
@@ -52,6 +52,7 @@ impl CCompilerImpl for GCC {
         cwd: &Path,
         env_vars: &[(OsString, OsString)],
         may_dist: bool,
+        rewrite_includes_only: bool,
     ) -> SFuture<process::Output>
     where
         T: CommandCreatorSync,
@@ -64,6 +65,7 @@ impl CCompilerImpl for GCC {
             env_vars,
             may_dist,
             self.kind(),
+            rewrite_includes_only,
         )
     }
 
@@ -74,8 +76,16 @@ impl CCompilerImpl for GCC {
         parsed_args: &ParsedArguments,
         cwd: &Path,
         env_vars: &[(OsString, OsString)],
+        rewrite_includes_only: bool,
     ) -> Result<(CompileCommand, Option<dist::CompileCommand>, Cacheable)> {
-        generate_compile_commands(path_transformer, executable, parsed_args, cwd, env_vars)
+        generate_compile_commands(
+            path_transformer,
+            executable,
+            parsed_args,
+            cwd,
+            env_vars,
+            rewrite_includes_only,
+        )
     }
 }
 
@@ -445,6 +455,7 @@ pub fn preprocess<T>(
     env_vars: &[(OsString, OsString)],
     may_dist: bool,
     kind: CCompilerKind,
+    rewrite_includes_only: bool,
 ) -> SFuture<process::Output>
 where
     T: CommandCreatorSync,
@@ -465,14 +476,16 @@ where
     if !may_dist && !parsed_args.profile_generate {
         cmd.arg("-P");
     }
-    match kind {
-        CCompilerKind::Clang => {
-            cmd.arg("-frewrite-includes");
+    if rewrite_includes_only {
+        match kind {
+            CCompilerKind::Clang => {
+                cmd.arg("-frewrite-includes");
+            }
+            CCompilerKind::GCC => {
+                cmd.arg("-fdirectives-only");
+            }
+            _ => (),
         }
-        CCompilerKind::GCC => {
-            cmd.arg("-fdirectives-only");
-        }
-        _ => (),
     }
     cmd.arg(&parsed_args.input)
         .args(&parsed_args.preprocessor_args)
@@ -493,6 +506,7 @@ pub fn generate_compile_commands(
     parsed_args: &ParsedArguments,
     cwd: &Path,
     env_vars: &[(OsString, OsString)],
+    rewrite_includes_only: bool,
 ) -> Result<(CompileCommand, Option<dist::CompileCommand>, Cacheable)> {
     #[cfg(not(feature = "dist-client"))]
     let _ = path_transformer;
@@ -534,12 +548,16 @@ pub fn generate_compile_commands(
     #[cfg(feature = "dist-client")]
     let dist_command = (|| {
         // https://gcc.gnu.org/onlinedocs/gcc-4.9.0/gcc/Overall-Options.html
-        let language = match parsed_args.language {
+        let mut language: String = match parsed_args.language {
             Language::C => "c",
             Language::Cxx => "c++",
             Language::ObjectiveC => "objective-c",
             Language::ObjectiveCxx => "objective-c++",
-        };
+        }
+        .into();
+        if !rewrite_includes_only {
+            language.push_str("-cpp-output");
+        }
         let mut arguments: Vec<String> = vec![
             "-x".into(),
             language.into(),
@@ -1150,6 +1168,7 @@ mod test {
             &parsed_args,
             f.tempdir.path(),
             &[],
+            false,
         )
         .unwrap();
         #[cfg(feature = "dist-client")]

--- a/src/compiler/gcc.rs
+++ b/src/compiler/gcc.rs
@@ -56,7 +56,15 @@ impl CCompilerImpl for GCC {
     where
         T: CommandCreatorSync,
     {
-        preprocess(creator, executable, parsed_args, cwd, env_vars, may_dist)
+        preprocess(
+            creator,
+            executable,
+            parsed_args,
+            cwd,
+            env_vars,
+            may_dist,
+            self.kind(),
+        )
     }
 
     fn generate_compile_commands(
@@ -113,7 +121,7 @@ counted_array!(pub static ARGS: [ArgInfo<ArgData>; _] = [
     take_arg!("--sysroot", PathBuf, Separated, PassThroughPath),
     take_arg!("-A", OsString, Separated, PassThrough),
     take_arg!("-B", PathBuf, CanBeSeparated, PassThroughPath),
-    take_arg!("-D", OsString, CanBeSeparated, PreprocessorArgument),
+    take_arg!("-D", OsString, CanBeSeparated, PassThrough),
     flag!("-E", TooHardFlag),
     take_arg!("-F", PathBuf, CanBeSeparated, PreprocessorArgumentPath),
     take_arg!("-G", OsString, Separated, PassThrough),
@@ -128,7 +136,7 @@ counted_array!(pub static ARGS: [ArgInfo<ArgData>; _] = [
     take_arg!("-MQ", OsString, Separated, PreprocessorArgument),
     take_arg!("-MT", OsString, Separated, DepTarget),
     flag!("-P", TooHardFlag),
-    take_arg!("-U", OsString, CanBeSeparated, PreprocessorArgument),
+    take_arg!("-U", OsString, CanBeSeparated, PassThrough),
     take_arg!("-V", OsString, Separated, PassThrough),
     take_arg!("-Xassembler", OsString, Separated, PassThrough),
     take_arg!("-Xlinker", OsString, Separated, PassThrough),
@@ -436,6 +444,7 @@ pub fn preprocess<T>(
     cwd: &Path,
     env_vars: &[(OsString, OsString)],
     may_dist: bool,
+    kind: CCompilerKind,
 ) -> SFuture<process::Output>
 where
     T: CommandCreatorSync,
@@ -455,6 +464,15 @@ where
     // With -fprofile-generate line number information is important, so don't use -P.
     if !may_dist && !parsed_args.profile_generate {
         cmd.arg("-P");
+    }
+    match kind {
+        CCompilerKind::Clang => {
+            cmd.arg("-frewrite-includes");
+        }
+        CCompilerKind::GCC => {
+            cmd.arg("-fdirectives-only");
+        }
+        _ => (),
     }
     cmd.arg(&parsed_args.input)
         .args(&parsed_args.preprocessor_args)
@@ -517,10 +535,10 @@ pub fn generate_compile_commands(
     let dist_command = (|| {
         // https://gcc.gnu.org/onlinedocs/gcc-4.9.0/gcc/Overall-Options.html
         let language = match parsed_args.language {
-            Language::C => "cpp-output",
-            Language::Cxx => "c++-cpp-output",
-            Language::ObjectiveC => "objective-c-cpp-output",
-            Language::ObjectiveCxx => "objective-c++-cpp-output",
+            Language::C => "c",
+            Language::Cxx => "c++",
+            Language::ObjectiveC => "objective-c",
+            Language::ObjectiveCxx => "objective-c++",
         };
         let mut arguments: Vec<String> = vec![
             "-x".into(),
@@ -530,7 +548,6 @@ pub fn generate_compile_commands(
             "-o".into(),
             path_transformer.to_dist(out_file)?,
         ];
-        // We could do preprocessor_args here, but skip for consistency with msvc
         arguments.extend(dist::osstrings_to_strings(&parsed_args.common_args)?);
         Some(dist::CompileCommand {
             executable: path_transformer.to_dist(&executable)?,

--- a/src/compiler/msvc.rs
+++ b/src/compiler/msvc.rs
@@ -63,6 +63,7 @@ impl CCompilerImpl for MSVC {
         cwd: &Path,
         env_vars: &[(OsString, OsString)],
         may_dist: bool,
+        _rewrite_includes_only: bool,
     ) -> SFuture<process::Output>
     where
         T: CommandCreatorSync,
@@ -85,6 +86,7 @@ impl CCompilerImpl for MSVC {
         parsed_args: &ParsedArguments,
         cwd: &Path,
         env_vars: &[(OsString, OsString)],
+        _rewrite_includes_only: bool,
     ) -> Result<(CompileCommand, Option<dist::CompileCommand>, Cacheable)> {
         generate_compile_commands(path_transformer, executable, parsed_args, cwd, env_vars)
     }

--- a/src/compiler/rust.rs
+++ b/src/compiler/rust.rs
@@ -997,6 +997,7 @@ where
         env_vars: Vec<(OsString, OsString)>,
         _may_dist: bool,
         pool: &CpuPool,
+        _rewrite_includes_only: bool,
     ) -> SFuture<HashResult> {
         let me = *self;
         #[rustfmt::skip] // https://github.com/rust-lang/rustfmt/issues/3759
@@ -1260,6 +1261,7 @@ impl Compilation for RustCompilation {
     fn generate_compile_commands(
         &self,
         path_transformer: &mut dist::PathTransformer,
+        _rewrite_includes_only: bool,
     ) -> Result<(CompileCommand, Option<dist::CompileCommand>, Cacheable)> {
         let RustCompilation {
             ref executable,
@@ -2692,6 +2694,7 @@ c:/foo/bar.rs:
                 .to_vec(),
                 false,
                 &pool,
+                false,
             )
             .wait()
             .unwrap();
@@ -2777,6 +2780,7 @@ c:/foo/bar.rs:
                 env_vars.to_owned(),
                 false,
                 &pool,
+                false,
             )
             .wait()
             .unwrap()

--- a/src/config.rs
+++ b/src/config.rs
@@ -383,6 +383,7 @@ pub struct DistConfig {
     pub cache_dir: PathBuf,
     pub toolchains: Vec<DistToolchainConfig>,
     pub toolchain_cache_size: u64,
+    pub rewrite_includes_only: bool,
 }
 
 impl Default for DistConfig {
@@ -393,6 +394,7 @@ impl Default for DistConfig {
             cache_dir: default_dist_cache_dir(),
             toolchains: Default::default(),
             toolchain_cache_size: default_toolchain_cache_size(),
+            rewrite_includes_only: true,
         }
     }
 }

--- a/src/server.rs
+++ b/src/server.rs
@@ -147,6 +147,7 @@ struct DistClientConfig {
     cache_dir: PathBuf,
     toolchain_cache_size: u64,
     toolchains: Vec<config::DistToolchainConfig>,
+    rewrite_includes_only: bool,
 }
 
 #[cfg(feature = "dist-client")]
@@ -196,6 +197,7 @@ impl DistClientContainer {
             cache_dir: config.dist.cache_dir.clone(),
             toolchain_cache_size: config.dist.toolchain_cache_size,
             toolchains: config.dist.toolchains.clone(),
+            rewrite_includes_only: config.dist.rewrite_includes_only,
         };
         let state = Self::create_state(config);
         Self {
@@ -348,6 +350,7 @@ impl DistClientContainer {
                     config.toolchain_cache_size,
                     &config.toolchains,
                     auth_token,
+                    config.rewrite_includes_only,
                 );
                 let dist_client = try_or_retry_later!(
                     dist_client.chain_err(|| "failure during dist client creation")

--- a/tests/dist.rs
+++ b/tests/dist.rs
@@ -35,11 +35,12 @@ fn basic_compile(tmpdir: &Path, sccache_cfg_path: &Path, sccache_cached_cfg_path
     ];
     let source_file = "x.c";
     let obj_file = "x.o";
-    write_source(tmpdir, source_file, "int x() { return 5; }");
+    write_source(tmpdir, source_file, "#if !defined(SCCACHE_TEST_DEFINE)\n#error SCCACHE_TEST_DEFINE is not defined\n#endif\nint x() { return 5; }");
     sccache_command()
         .args(&[
             std::env::var("CC").unwrap_or("gcc".to_string()).as_str(),
             "-c",
+            "-DSCCACHE_TEST_DEFINE",
         ])
         .arg(tmpdir.join(source_file))
         .arg("-o")

--- a/tests/harness/mod.rs
+++ b/tests/harness/mod.rs
@@ -261,6 +261,8 @@ impl DistSystem {
         let mut child = Command::new("docker")
             .args(&["build", "-q", "-t", DIST_IMAGE, "-"])
             .stdin(Stdio::piped())
+            .stdout(Stdio::piped())
+            .stderr(Stdio::piped())
             .spawn()
             .unwrap();
         child
@@ -667,8 +669,8 @@ fn make_container_name(tag: &str) -> String {
 
 fn check_output(output: &Output) {
     if !output.status.success() {
-        println!("[BEGIN OUTPUT]\n===========\n{}\n==========\n\n\n\n=========\n{}\n===============\n[FIN OUTPUT]\n\n",
-            String::from_utf8_lossy(&output.stdout), String::from_utf8_lossy(&output.stderr));
+        println!("{}\n\n[BEGIN STDOUT]\n===========\n{}\n===========\n[FIN STDOUT]\n\n[BEGIN STDERR]\n===========\n{}\n===========\n[FIN STDERR]\n\n",
+            output.status, String::from_utf8_lossy(&output.stdout), String::from_utf8_lossy(&output.stderr));
         panic!()
     }
 }

--- a/tests/harness/mod.rs
+++ b/tests/harness/mod.rs
@@ -194,6 +194,7 @@ pub fn sccache_client_cfg(tmpdir: &Path) -> sccache::config::FileConfig {
             cache_dir: tmpdir.join(dist_cache_relpath),
             toolchains: vec![],
             toolchain_cache_size: TC_CACHE_SIZE,
+            rewrite_includes_only: false, // TODO
         },
     }
 }

--- a/tests/oauth.rs
+++ b/tests/oauth.rs
@@ -59,6 +59,7 @@ fn config_with_dist_auth(
             cache_dir: tmpdir.join("unused-cache"),
             toolchains: vec![],
             toolchain_cache_size: 0,
+            rewrite_includes_only: true,
         },
     }
 }

--- a/tests/system.rs
+++ b/tests/system.rs
@@ -64,6 +64,8 @@ macro_rules! vec_from {
     };
 }
 
+// TODO: This will fail if gcc/clang is actually a ccache wrapper, as it is the
+// default case on Fedora, e.g.
 fn compile_cmdline<T: AsRef<OsStr>>(
     compiler: &str,
     exe: T,

--- a/tests/system.rs
+++ b/tests/system.rs
@@ -81,6 +81,16 @@ const INPUT: &'static str = "test.c";
 const INPUT_ERR: &'static str = "test_err.c";
 const OUTPUT: &'static str = "test.o";
 
+// Copy the source files into the tempdir so we can compile with relative paths, since the commandline winds up in the hash key.
+fn copy_to_tempdir(inputs: &[&str], tempdir: &Path) {
+    for f in inputs {
+        let original_source_file = Path::new(file!()).parent().unwrap().join(f.clone());
+        let source_file = tempdir.join(f);
+        trace!("fs::copy({:?}, {:?})", original_source_file, source_file);
+        fs::copy(&original_source_file, &source_file).unwrap();
+    }
+}
+
 fn test_basic_compile(compiler: Compiler, tempdir: &Path) {
     let Compiler {
         name,
@@ -89,13 +99,7 @@ fn test_basic_compile(compiler: Compiler, tempdir: &Path) {
     } = compiler;
     trace!("run_sccache_command_test: {}", name);
     // Compile a source file.
-    // Copy the source files into the tempdir so we can compile with relative paths, since the commandline winds up in the hash key.
-    for f in &[INPUT, INPUT_ERR] {
-        let original_source_file = Path::new(file!()).parent().unwrap().join(f);
-        let source_file = tempdir.join(f);
-        trace!("fs::copy({:?}, {:?})", original_source_file, source_file);
-        fs::copy(&original_source_file, &source_file).unwrap();
-    }
+    copy_to_tempdir(&[INPUT, INPUT_ERR], tempdir);
 
     let out_file = tempdir.join("test.o");
     trace!("compile");
@@ -151,11 +155,7 @@ fn test_noncacheable_stats(compiler: Compiler, tempdir: &Path) {
         env_vars,
     } = compiler;
     trace!("test_noncacheable_stats: {}", name);
-    // Copy the source file into the tempdir so we can compile with relative paths, since the commandline winds up in the hash key.
-    let original_source_file = Path::new(file!()).parent().unwrap().join(INPUT);
-    let source_file = tempdir.join(INPUT);
-    trace!("fs::copy({:?}, {:?})", original_source_file, source_file);
-    fs::copy(&original_source_file, &source_file).unwrap();
+    copy_to_tempdir(&[INPUT], tempdir);
 
     trace!("compile");
     Command::main_binary()

--- a/tests/system.rs
+++ b/tests/system.rs
@@ -390,6 +390,10 @@ fn find_compilers() -> Vec<Compiler> {
     }]
 }
 
+// TODO: This runs multiple test cases, for multiple compilers. It should be
+// split up to run them individually. In the current form, it is hard to see
+// which sub test cases are executed, and if one fails, the remaining tests
+// are not run.
 #[test]
 #[cfg(any(unix, target_env = "msvc"))]
 fn test_sccache_command() {

--- a/tests/system.rs
+++ b/tests/system.rs
@@ -82,6 +82,7 @@ fn compile_cmdline<T: AsRef<OsStr>>(
 const INPUT: &'static str = "test.c";
 const INPUT_ERR: &'static str = "test_err.c";
 const INPUT_MACRO_EXPANSION: &'static str = "test_macro_expansion.c";
+const INPUT_WITH_DEFINE: &'static str = "test_with_define.c";
 const OUTPUT: &'static str = "test.o";
 
 // Copy the source files into the tempdir so we can compile with relative paths, since the commandline winds up in the hash key.
@@ -344,8 +345,35 @@ fn test_gcc_clang_no_warnings_from_macro_expansion(compiler: Compiler, tempdir: 
         .stderr(predicates::str::contains("warning:").from_utf8().not());
 }
 
+fn test_compile_with_define(compiler: Compiler, tempdir: &Path) {
+    let Compiler {
+        name,
+        exe,
+        env_vars,
+    } = compiler;
+    trace!("test_compile_with_define: {}", name);
+    // Compile a source file.
+    copy_to_tempdir(&[INPUT_WITH_DEFINE], tempdir);
+
+    trace!("compile");
+    sccache_command()
+        .args(
+            [
+                &compile_cmdline(name, &exe, INPUT_WITH_DEFINE, OUTPUT)[..],
+                &vec_from!(OsString, "-DSCCACHE_TEST_DEFINE")[..],
+            ]
+            .concat(),
+        )
+        .current_dir(tempdir)
+        .envs(env_vars.clone())
+        .assert()
+        .success()
+        .stderr(predicates::str::contains("warning:").from_utf8().not());
+}
+
 fn run_sccache_command_tests(compiler: Compiler, tempdir: &Path) {
     test_basic_compile(compiler.clone(), tempdir);
+    test_compile_with_define(compiler.clone(), tempdir);
     if compiler.name == "cl.exe" {
         test_msvc_deps(compiler.clone(), tempdir);
     }

--- a/tests/system.rs
+++ b/tests/system.rs
@@ -79,6 +79,7 @@ fn compile_cmdline<T: AsRef<OsStr>>(
 
 const INPUT: &'static str = "test.c";
 const INPUT_ERR: &'static str = "test_err.c";
+const INPUT_MACRO_EXPANSION: &'static str = "test_macro_expansion.c";
 const OUTPUT: &'static str = "test.o";
 
 // Copy the source files into the tempdir so we can compile with relative paths, since the commandline winds up in the hash key.
@@ -315,6 +316,32 @@ int main(int argc, char** argv) {
     });
 }
 
+fn test_gcc_clang_no_warnings_from_macro_expansion(compiler: Compiler, tempdir: &Path) {
+    let Compiler {
+        name,
+        exe,
+        env_vars,
+    } = compiler;
+    trace!("test_gcc_clang_no_warnings_from_macro_expansion: {}", name);
+    // Compile a source file.
+    copy_to_tempdir(&[INPUT_MACRO_EXPANSION], tempdir);
+
+    trace!("compile");
+    sccache_command()
+        .args(
+            [
+                &compile_cmdline(name, &exe, INPUT_MACRO_EXPANSION, OUTPUT)[..],
+                &vec_from!(OsString, "-Wunreachable-code")[..],
+            ]
+            .concat(),
+        )
+        .current_dir(tempdir)
+        .envs(env_vars.clone())
+        .assert()
+        .success()
+        .stderr(predicates::str::contains("warning:").from_utf8().not());
+}
+
 fn run_sccache_command_tests(compiler: Compiler, tempdir: &Path) {
     test_basic_compile(compiler.clone(), tempdir);
     if compiler.name == "cl.exe" {
@@ -323,6 +350,9 @@ fn run_sccache_command_tests(compiler: Compiler, tempdir: &Path) {
     if compiler.name == "gcc" {
         test_gcc_mp_werror(compiler.clone(), tempdir);
         test_gcc_fprofile_generate_source_changes(compiler.clone(), tempdir);
+    }
+    if compiler.name == "clang" || compiler.name == "gcc" {
+        test_gcc_clang_no_warnings_from_macro_expansion(compiler.clone(), tempdir);
     }
 }
 

--- a/tests/test_macro_expansion.c
+++ b/tests/test_macro_expansion.c
@@ -1,0 +1,10 @@
+#include <stdlib.h>
+
+#define foo(x)                                                                 \
+  {                                                                            \
+    if (x) {                                                                   \
+      abort();                                                                 \
+    }                                                                          \
+  }
+
+void bar() { foo(0); }

--- a/tests/test_with_define.c
+++ b/tests/test_with_define.c
@@ -1,0 +1,7 @@
+#include <stdio.h>
+
+#if !defined(SCCACHE_TEST_DEFINE)
+#error SCCACHE_TEST_DEFINE is not defined
+#endif
+
+void foo() { printf("hello world\n"); }

